### ISSUE TITLE
Estimators Siamese Model Peformance Issue 

### ISSUE THREAD

**Issue by imranparuk** [tensorflow/tensorflow#23136]
tf.esitmators producing poor performance vs TF & Keras.  

* System information
    OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
    TensorFlow installed from (source or binary):
    pip3 install --upgrade tensorflow-gpu==1.12.0-rc1
    TensorFlow version: 1.12.0-rc1
    Python version: Python 3.6.5 
    Installed using virtualenv? pip? conda?: pip3
    CUDA/cuDNN version: 9.0 / 7.0.5
    GPU model and memory: nvidia gtx 1050 (Lenovo Laptop)

* Code to reproduce performance issue:

```python
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

from keras.datasets import mnist

import numpy as np
import tensorflow as tf

import random
tf.logging.set_verbosity(tf.logging.INFO)


def cnn_cnn_model(features, reuse=False):
    """Model function for CNN."""

    # Flat 1
    flat_1 = tf.layers.flatten(
        features,
        name="flattern",
    )

    # dense 1
    dense_1 = tf.layers.dense(
        flat_1,
        1024,
        activation=tf.nn.relu,
        reuse=reuse,
        name='dense_1',
    )

    # drop 1
    drop_1 = tf.layers.dropout(
        dense_1,
        rate=0.1,
        name="drop_1",
    )

    # dense 2
    dense_2 = tf.layers.dense(
        drop_1,
        512,
        activation=tf.nn.relu,
        reuse=reuse,
        name='dense_2',
    )

    # drop 2
    drop_2 = tf.layers.dropout(
        dense_2,
        rate=0.1,
        name="drop_2",
    )

    # dense 3
    dense_3 = tf.layers.dense(
        drop_2,
        128,
        activation=None,
        reuse=reuse,
        name='dense_3',
    )

    return dense_3


def accuracy(y_true, y_pred):
    '''Compute classification accuracy with a fixed threshold on distances.
    '''

    return tf.metrics.mean(tf.equal(y_true, tf.cast(y_pred < tf.cast(0.5, 'float64'), y_true.dtype)))

def compute_accuracy(y_true, y_pred):
    '''Compute classification accuracy with a fixed threshold on distances.
    '''
    pred = y_pred.ravel() < 0.5
    return np.mean(pred == y_true)


def my_loss(dist, y_true, margin=1.0):

    y_true = tf.cast(y_true, 'float64')
    margin = tf.cast(margin, 'float64')
    dist = tf.cast(dist, 'float64')

    # Loss function
    loss_pos = tf.multiply(y_true, tf.pow(dist, 2), name='constrastive_loss_1')
    loss_neg = tf.multiply(tf.subtract(tf.cast(1.0, 'float64'), y_true),
                           tf.pow(tf.maximum(tf.subtract(margin, dist), 0), 2),
                           name='constrastive_loss_2')
    loss = tf.reduce_mean(tf.add(loss_neg, loss_pos), name='constrastive_loss')

    return loss


def euclidean_distance(x1, x2):
    x1 = tf.cast(x1, 'float64')
    x2 = tf.cast(x2, 'float64')
    epsilon = tf.cast(1e-7, 'float64')

    eucd2 = tf.pow(tf.subtract(x1, x2), 2, name='eucd2')
    eucd2 = tf.reduce_sum(eucd2, 1)

    eucd2 = tf.maximum(eucd2, epsilon)
    eucd = tf.sqrt(eucd2, name='eucd')

    return eucd



def create_siamese(features, labels, mode):
    # Input Layers
    input_layer_a = tf.reshape(features["x_1"], [-1, 28, 28])
    input_layer_b = tf.reshape(features["x_2"], [-1, 28, 28])

    with tf.variable_scope("siamese") as scope:

        network_a = cnn_cnn_model(input_layer_a, reuse=False)
        network_b = cnn_cnn_model(input_layer_b, reuse=True)

    if mode == tf.estimator.ModeKeys.TRAIN:
        dist = euclidean_distance(network_a, network_b)
        loss = my_loss(dist, labels)

        optimizer = tf.train.AdamOptimizer(learning_rate=0.001)
        train_op = optimizer.minimize(
            loss=loss,
            global_step=tf.train.get_global_step())
        return tf.estimator.EstimatorSpec(mode=mode, loss=loss, train_op=train_op)

    elif mode == tf.estimator.ModeKeys.PREDICT:
        predictions = {
            "eucl": euclidean_distance(network_a, network_b)
        }

        return tf.estimator.EstimatorSpec(mode=mode, predictions=predictions)


def create_pairs(x, digit_indices, num_classes):
    '''Positive and negative pair creation.
    Alternates between positive and negative pairs.
    '''
    pairs = []
    labels = []
    n = min([len(digit_indices[d]) for d in range(num_classes)]) - 1
    for d in range(num_classes):
        for i in range(n):
            z1, z2 = digit_indices[d][i], digit_indices[d][i + 1]
            pairs += [[x[z1], x[z2]]]
            inc = random.randrange(1, num_classes)
            dn = (d + inc) % num_classes
            z1, z2 = digit_indices[d][i], digit_indices[dn][i]
            pairs += [[x[z1], x[z2]]]
            labels += [1, 0]
    return np.array(pairs), np.array(labels)

def main(unused_argv):

    path_ = '/media/imran/bigboy/datasets/my_smallVoxCeleb/updated_datasets/split_datasets/3d_cnn/smallest_dataset.hdf5'
    num_classes = 10

    # the data, split between train and test sets
    (x_train, y_train), (x_test, y_test) = mnist.load_data()
    x_train = x_train.astype('float64')
    x_test = x_test.astype('float64')
    y_train = y_train.astype('float64')
    y_test = y_test.astype('float64')


    x_train /= 255
    x_test /= 255

    # create training+test positive and negative pairs
    digit_indices = [np.where(y_train == i)[0] for i in range(num_classes)]
    tr_pairs, tr_y = create_pairs(x_train, digit_indices, num_classes)

    digit_indices = [np.where(y_test == i)[0] for i in range(num_classes)]
    te_pairs, te_y = create_pairs(x_test, digit_indices, num_classes)


    # Create the Estimator
    mnist_classifier = tf.estimator.Estimator(
      model_fn=create_siamese)

    # Train the model
    train_input_fn = tf.estimator.inputs.numpy_input_fn(
      x={"x_1": tr_pairs[:, 0],
         "x_2": tr_pairs[:, 1]},
      y=tr_y,
      batch_size=128,
      num_epochs=1,
      shuffle=True)

    test_input_fn_train = tf.estimator.inputs.numpy_input_fn(
        x={"x_1": tr_pairs[:, 0],
           "x_2": tr_pairs[:, 1]},
        shuffle=False
       )

    test_input_fn_test = tf.estimator.inputs.numpy_input_fn(
        x={"x_1": te_pairs[:, 0],
           "x_2": te_pairs[:, 1]},
        shuffle=False
       )

    mnist_classifier.train(
        input_fn=train_input_fn,
        steps=None)

    predictions_test = list(mnist_classifier.predict(input_fn=test_input_fn_test))
    predictions_train = list(mnist_classifier.predict(input_fn=test_input_fn_train))

    preds_test = []
    for item in predictions_test:
        preds_test.append(item['eucl'])

    preds_train = []
    for item in predictions_train:
        preds_train.append(item['eucl'])

    print("Accuracy on train set: ", compute_accuracy(tr_y, np.array(preds_train)))
    print("Accuracy on test set: ", compute_accuracy(te_y, np.array(preds_test)))


if __name__ == "__main__":
    tf.app.run()
```

With Keras model.fit(), this yields:
```
Accuracy on training set: 95.70%
Accuracy on test set: 95.33%
```
With TF this yields:
```
Accuracy on training set: 94.56%
Accuracy on test set: 94.53%
```
With tf.estimators this yields:
```
Accuracy on training set: 61.32%
Accuracy on test set: 62.74%
```

The code for the benchmark code for Keras and TF can be found here:
+ Keras:
https://gist.github.com/imranparuk/25963eb3b2db1540ead684271de6f5a8
+ TF:
https://gist.github.com/imranparuk/4fe48323a006ff030bf2037136db7868

**Comment by bhack** [comment]
Have you tested with Model_to_estimator?

**Comment by imranparuk** [comment]
@tensorflowbutler 
+ Have I written custom code: Yes, provide above
+ Bazel version: N/A
+ Exact command to reproduce: N/A
+ Mobile device: N/A

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ymodak: It has been 44 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
Apologies for the delay in response. Is this still an issue?
In order to expedite the trouble-shooting process, please provide a minimal code snippet to reproduce the issue reported here. Thanks!

**Comment by imranparuk** [comment]
I've moved my code to pytorch, so yeah my issue is solved.
